traits/has_traits.py

```diff
@@ -115,6 +115,30 @@ class AbstractViewElement(abc.ABC):
 # Generic 'Any' trait:
 any_trait = Any().as_ctrait()
 
+# Classes where we don't check for name collisions.
+HasTraitsSubclasses = (
+    "HasTraits",
+    "HasStrictTraits",
+    "HasRequiredTraits",
+    "HasPrivateTraits",
+    "ABCMetaHasTraits",
+    "ABCHasTraits",
+    "ABCHasStrictTraits",
+    "SingletonHasTraits",
+    "SingletonHasStrictTraits",
+    "SingletonHasPrivateTraits",
+    "Vetoable",
+    "MetaInterface",
+    "Interface",
+    "ISerializable",
+    "traits_super"
+)
+
+
+# A custom warning issued by Traits
+class TraitsWarning(Warning):
+    pass
+
```

Comment:
I think we would want to expose this class on `api.py` so that downstream projects can import it for capturing the said warning.
Please add a class docstring to it as well.  We will also need to update the documentation to mention this class.

Reply:
Is this the right place to define it ? There is a `trait_errors` module for errors but nothing equivalent for warnings . Not sure if I should create a new module, or maybe put it in `trait_errors`?

Reply:
Why not use one of the built-in warning classes rather than defining our own? The warnings module provides good functionality for filtering by module or by message, so I'm not sure we need a dedicated `Warning` subclass here.

Reply:
Specifically, I think `UserWarning` would work here. Alternatively, we might consider `DeprecationWarning`: if we go the route of carving off a namespace for Traits' sole use, then we'll eventually want to make used-defined trait names starting with `_trait` or `trait` an error. The only reason not to raise right now is backwards compatibility.

I think the problem with `TraitsWarning` is that it's not clear what it's actually for. If the plan is to use it for any warning issued by Traits now or in the future, then I think that's a bad idea: I think we want our warning type to be related to the type of condition that's being warned about, rather than related to a particular package - we already have module-level filtering for that.

If the plan is to use it for this _specific_ name clash issue, then it probably needs a more descriptive name. And then we risk ending up with various warnings in the future. It seems simpler all round to use the built-in warning classes unless we have a good reason not to.



I see a problem with `TraitsWarning`

I also think the `TraitError` / `TraitsWarning` name mismatch is going to be confusing.

Reply:
Agreed that the good old `UserWarning` would also do the job here.

Reply:
My thinking was that it would be easier to filter out `TraitsWarning` especially when, for example ,analyzing a log file of some long running application, but I guess the module level filtering covers it and the warning should be more specific. Anyway, I understand why it's better to not have it. Removed.